node/src/components/small_network/bandwidth_limiter.rs

```diff
@@ -0,0 +1,429 @@
+//! Bandwidth limiters
+//!
+//! Bandwidth limiters restrict the usable amount of bandwidth through slowing down the sending of
+//! message by making each sender request an allowance for sending first.
+
+use std::{collections::HashSet, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use casper_types::PublicKey;
+use tokio::{
+    sync::{mpsc, oneshot},
+    time::Instant,
+};
+use tracing::{debug, warn};
+
+use crate::types::NodeId;
+
+/// Amount of allowed bandwidth to buffer in `ClassBasedLimiter`.
+const STORED_BUFFER_SECS: Duration = Duration::from_secs(2);
+
+/// A bandwidth limiter.
+///
+/// Any sender for a specific connection is expected to call `create_handle` for a connection, and
+/// using that handle to request a bandwidth allowance.
+pub(crate) trait BandwidthLimiter: Send + Sync {
+    /// Create a handle for a connection using the given peer and optional validator id.
+    fn create_handle(
+        &self,
+        peer_id: NodeId,
+        validator_id: Option<PublicKey>,
+    ) -> Box<dyn BandwidthLimiterHandle>;
+
+    /// Update the validator sets.
+    fn update_validators(
+        &self,
+        active_validators: HashSet<PublicKey>,
+        upcoming_validators: HashSet<PublicKey>,
+    );
+}
+
+/// A handle for connections that are bandwidth limited.
+#[async_trait]
+pub(crate) trait BandwidthLimiterHandle: Send + Sync {
+    /// Waits until the sender is clear to send `num_bytes` additional bytes.
+    async fn request_allowance(&self, num_bytes: u32);
+}
+
+/// An unlimited bandwidth "limiter".
+///
+/// Does not restrict outgoing bandwidth in any way (`create_handle` returns immediately).
```

Comment:
Was that supposed to say that `request_allowance` returns immediately?

Reply:
Yes. I'll correct that.